casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -276,41 +276,103 @@ class RuntimeManagerImpl[F[_]: Concurrent] private[rholang] (
     )
 
   private def newEval(
-      terms: Seq[DeployData],
-      runtime: Runtime[F],
-      initHash: StateHash
-  ): F[(StateHash, Seq[InternalProcessedDeploy])] = {
-
-    def doEval(
-        terms: Seq[DeployData],
-        hash: Blake2b256Hash,
-        acc: Seq[InternalProcessedDeploy]
-    ): F[(StateHash, Seq[InternalProcessedDeploy])] =
-      Concurrent[F].defer {
-        terms match {
-          case deploy +: rem =>
-            for {
-              evaluateResult               <- computeEffect(runtime)(deploy)
-              EvaluateResult(cost, errors) = evaluateResult
-              newCheckpoint                <- runtime.space.createCheckpoint()
-              deployResult = InternalProcessedDeploy(
-                deploy,
-                Cost.toProto(cost),
-                newCheckpoint.log,
-                Seq.empty[trace.Event],
-                DeployStatus.fromErrors(errors)
-              )
-              cont <- if (errors.isEmpty)
-                       doEval(rem, newCheckpoint.root, acc :+ deployResult)
-                     else doEval(rem, hash, acc :+ deployResult)
-            } yield cont
-
-          case _ => (ByteString.copyFrom(hash.bytes.toArray), acc).pure[F]
+      runtime: Runtime[F]
+  )(deploys: Seq[DeployData], withPayment: Boolean): F[(StateHash, Seq[InternalProcessedDeploy])] =
+    Concurrent[F]
+      .defer {
+        runtime.space.createCheckpoint() >>= { checkpoint0 =>
+          deploys.toList.foldM((checkpoint0.root, Seq.empty[InternalProcessedDeploy])) {
+            case ((prev, processedDeploys), deploy) =>
+              import deploy._
+              if (withPayment) {
+                computeBalance(runtime)(deployer) >>= { balance =>
+                  runtime.space.reset(prev) *> {
+                    if (balance >= phloLimit * phloPrice) {
+                      computeEffect(runtime)(deploy) >>= {
+                        case EvaluateResult(cost, errors) =>
+                          errors.isEmpty
+                            .pure[F]
+                            .ifM(
+                              ifTrue = ()
+                                .pure[F],
+                              ifFalse = runtime.space.reset(prev)
+                            ) *> {
+                            runtime.space.createCheckpoint() >>= { checkpoint1 =>
```

Comment:
Maybe name this `deployEvalCheckpoint` and the below one `deployPaymentCheckpoint`